source/server/connection_handler_impl.cc

```diff
@@ -293,54 +307,92 @@ void ConnectionHandlerImpl::ActiveTcpListener::newConnection(
-  new_connection->setBufferLimits(config_.perConnectionBufferLimitBytes());
+  ActiveTcpConnectionPtr active_connection(new ActiveTcpConnection(
+      *this,
+      parent_.dispatcher_.createServerConnection(std::move(socket), std::move(transport_socket)),
+      parent_.dispatcher_.timeSource()));
+  active_connection->connection_->setBufferLimits(config_.perConnectionBufferLimitBytes());
 
   const bool empty_filter_chain = !config_.filterChainFactory().createNetworkFilterChain(
-      *new_connection, filter_chain->networkFilterFactories());
+      *active_connection->connection_, filter_chain->networkFilterFactories());
   if (empty_filter_chain) {
-    ENVOY_CONN_LOG(debug, "closing connection: no filters", *new_connection);
-    new_connection->close(Network::ConnectionCloseType::NoFlush);
-    return;
+    ENVOY_CONN_LOG(debug, "closing connection: no filters", *active_connection->connection_);
+    active_connection->connection_->close(Network::ConnectionCloseType::NoFlush);
   }
 
-  onNewConnection(std::move(new_connection));
-}
-
-void ConnectionHandlerImpl::ActiveTcpListener::onNewConnection(
-    Network::ConnectionPtr&& new_connection) {
-  ENVOY_CONN_LOG(debug, "new connection", *new_connection);
-
   // If the connection is already closed, we can just let this connection immediately die.
-  if (new_connection->state() != Network::Connection::State::Closed) {
-    ActiveConnectionPtr active_connection(
-        new ActiveConnection(*this, std::move(new_connection), parent_.dispatcher_.timeSource()));
+  if (active_connection->connection_->state() != Network::Connection::State::Closed) {
+    ENVOY_CONN_LOG(debug, "new connection", *active_connection->connection_);
+    active_connection->connection_->addConnectionCallbacks(*active_connection);
     active_connection->moveIntoList(std::move(active_connection), connections_);
-    parent_.num_connections_++;
   }
 }
 
-ConnectionHandlerImpl::ActiveConnection::ActiveConnection(ActiveTcpListener& listener,
-                                                          Network::ConnectionPtr&& new_connection,
-                                                          TimeSource& time_source)
+namespace {
+// Structure used to allow a unique_ptr to be captured in a posted lambda. See below.
+struct RebalancedSocket {
+  Network::ConnectionSocketPtr socket;
+};
+using RebalancedSocketSharedPtr = std::shared_ptr<RebalancedSocket>;
+} // namespace
+
+void ConnectionHandlerImpl::ActiveTcpListener::post(Network::ConnectionSocketPtr&& socket) {
+  // It is not possible to capture a unique_ptr so we must bundle the socket inside a shared_ptr
+  // that can be captured.
+  RebalancedSocketSharedPtr socket_to_rebalance = std::make_shared<RebalancedSocket>();
+  socket_to_rebalance->socket = std::move(socket);
+
+  parent_.dispatcher_.post([socket_to_rebalance, tag = config_.listenerTag(), &parent = parent_]() {
```

Comment:
No, it's impossible to capture a unique_ptr in a lambda in C++ currently because lambdas are always copyable (unfortunately).

Reply:
it's supported by C++14 and upwards

Reply:
I've tried this multiple times and I'm quite certain it is not supported. Can you link me to something that says otherwise?

Reply:
I just searched online, and saw something like this: https://stackoverflow.com/questions/27800468/capture-and-move-a-unique-ptr-in-a-c14-lambda-expression/27800638.

Reply:
Sorry I remember what the issue is now. The issue is that the `post()` implementation itself copies the lambda. I think it might be possible to fix this but it would require changes to that interface and all callers. I tried to fix this a while ago and it turned into a giant PITA so I dropped it.

Reply:
I see. Thanks for your explanation! Can you comment on RebalancedSocket about this approach?

Reply:
Yup will do.